sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,327 @@
+    return new CassandraReaderImpl<>(source);
+  }
+
+  @Override
+  public long getEstimatedSizeBytes(CassandraIO.Read<T> spec) {
+    try (Cluster cluster = getCluster(spec.hosts(), spec.port())) {
+      if (isMurmur3Partitioner(cluster)) {
+        try {
+          List<TokenRange> tokenRanges = getTokenRanges(cluster,
+              spec.keyspace(),
+              spec.table());
+          return getEstimatedSizeBytes(tokenRanges);
+        } catch (Exception e) {
+          LOG.warn("Can't estimate the size", e);
+          return 0L;
+        }
+      } else {
+        return 0L;
+      }
+    }
+  }
+
+  /**
+   * Actually estimate the size of the data to read on the cluster, based on the given token
+   * ranges to address.
+   */
+  @VisibleForTesting
+  protected static long getEstimatedSizeBytes(List<TokenRange> tokenRanges) {
+    long size = 0L;
+    for (TokenRange tokenRange : tokenRanges) {
+      size = size + tokenRange.meanPartitionSize * tokenRange.partitionCount;
+    }
+    return Math.round(size / getRingFraction(tokenRanges));
+  }
+
+  @Override
+  public List<BoundedSource<T>> split(CassandraIO.Read<T> spec,
+      long desiredBundleSizeBytes) {
+    try (Cluster cluster = getCluster(spec.hosts(), spec.port())) {
+      if (isMurmur3Partitioner(cluster)) {
+        LOG.info("Murmur3Partitioner detected, splitting");
+        return split(spec, desiredBundleSizeBytes, getEstimatedSizeBytes(spec));
+      } else {
+        LOG.warn("Only Murmur3Partitioner is supported for splitting, using an unique source for "
+            + "the read");
+        String splitQuery = QueryBuilder.select().from(spec.keyspace(), spec.table()).toString();
+        List<BoundedSource<T>> sources = new ArrayList<>();
+        sources.add(new CassandraIO.CassandraSource<T>(spec, splitQuery));
+        return sources;
+      }
+    }
+  }
+
+  /**
+   * Compute the number of splits based on the estimated size and the desired bundle size, and
+   * create several sources.
+   */
+  @VisibleForTesting
+  protected List<BoundedSource<T>> split(CassandraIO.Read<T> spec,
+                                                long desiredBundleSizeBytes,
+                                                long estimatedSizeBytes) {
+    long numSplits = 1;
+    List<BoundedSource<T>> sourceList = new ArrayList<>();
+    if (desiredBundleSizeBytes > 0) {
+      numSplits = estimatedSizeBytes / desiredBundleSizeBytes;
+    }
+    if (numSplits <= 0) {
+      LOG.warn("Number of splits is less than 0 ({}), fallback to 1", numSplits);
+      numSplits = 1;
+    }
+
+    LOG.info("Number of splits is {}", numSplits);
+
+    double startRange = MIN_TOKEN;
+    double endRange = MAX_TOKEN;
+    double startToken, endToken;
+
+    endToken = startRange;
+    double incrementValue = endRange - startRange / numSplits;
+    String splitQuery;
+    for (int splitCount = 1; splitCount <= numSplits; splitCount++) {
```

Comment:
splitCount reads like "number of splits" - better to use a simple variable like "i" and range it from 0 to numSplits-1, like most indexed loops in Java do.